DISCUSSION-42: Upload of attachment is not properly supported

surli opened the pull request:
Provides change so that it's possible to know the reference of a message before it's actually created: this reference can then be used for upload attachments features, or resolution of relative links.

This commit provides the following changes:
  * Upgrade parent pom to XWiki 14.10 to be able to use the new APIs
  * Add a new interface MessageHolderReferenceService in store-api for being able to retrieve reference of a message and use it in the discussion script service
  * Refactor computation of references in store-default to rely on a single dedicated component, and to use UUID#randomString instead of a custom random implementation
  * Base default implementation of the MessageHolderReferenceService on the HttpSession so that the created MessageReferences are attached to current user session (inspired from TemporaryAttachmentSession)
  * Refactor on the way to display the message editor, and rename it
  * Minor improvment to check existence of the wiki in the handler

## TODO
  - [x] Perform manual tests
  - [x] Add some unit tests for new components

surli committed:
DISCUSSION-42: Upload of attachment is not properly supported

Provides change so that it's possible to know the reference of a message
before it's actually created: this reference can then be used for upload
attachments features, or resolution of relative links.

This commit provides the following changes:
  * Upgrade parent pom to XWiki 14.10 to be able to use the new APIs
  * Add a new interface MessageHolderReferenceService in store-api for being able to retrieve
    reference of a message and use it in the discussion script service
  * Refactor computation of references in store-default to rely on a
    single dedicated component, and to use UUID#randomString instead of
a custom random implementation
  * Base default implementation of the MessageHolderReferenceService on
    the HttpSession so that the created MessageReferences are attached
to current user session (inspired from TemporaryAttachmentSession)
  * Refactor on the way to display the message editor, and rename it
  * Minor improvment to check existence of the wiki in the handler

manuelleduc commented on discussions-store/discussions-store-default/src/main/java/org/xwiki/contrib/discussions/store/internal/DefaultMessageHolderReferenceService.java:
```diff
@@ -0,0 +1,114 @@
+@Singleton
+public class DefaultMessageHolderReferenceService implements MessageHolderReferenceService
+{
+    static final String MESSAGE_HOLDER_REFERENCE_SESSION_ATTRIBUTE = "xwikiDiscussionsMessageHolderReferences";
```
no visibility?

surli replied:
indeed, I could keep it strictly private

manuelleduc commented on discussions-store/discussions-store-default/src/main/java/org/xwiki/contrib/discussions/store/internal/PageHolderReferenceFactory.java:
```diff
@@ -0,0 +1,124 @@
+            this.discussionStoreConfigurationFactory.getDiscussionStoreConfiguration(applicationHint);
+
+        SpaceReference spaceLocation;
+        switch (entity) {
```
Not a fan of this closed enumeration. One alternative is to let one of the parameter be a lambda (probably a `Callable`), and to simply call it to initialize `spaceLocation`.

surli replied:
This is only internal and not supposed to evolve. Also I really don't want to put back retrieving the space location in the different services as it's really similar from one entity to another.

manuelleduc replied:
Sounds good, it's a net improvement anyway :)

manuelleduc commented on discussions-server/src/main/java/org/xwiki/contrib/discussions/internal/server/DiscussionsResourceReferenceHandler.java:
```diff
@@ -171,6 +177,7 @@ public void handle(ResourceReference reference, ResourceReferenceHandlerChain ch
             XWikiContext context = this.contextProvider.get();
             WikiReference currentWiki = null;
             if (discussionsResourceReference.getWikiReference() != null) {
+                this.checkWiki(discussionsResourceReference);
```
Very minor comment, method calls are not suppose to start with `this`.

manuelleduc commented on discussions-server/src/main/java/org/xwiki/contrib/discussions/internal/server/DiscussionsResourceReferenceHandler.java:
```diff
@@ -207,6 +214,23 @@ public void handle(ResourceReference reference, ResourceReferenceHandlerChain ch
+    private void checkWiki(DiscussionsResourceReference discussionsResourceReference)
+        throws ResourceReferenceHandlerException
+    {
+        if (discussionsResourceReference.getWikiReference() != null) {
```
No error when the wiki reference is missing?

surli replied:
I will double check but AFAIR we added later the support of wiki references there and so we accept it to be null for backward compatibility reasons.

surli replied:
You can actually see line 179 that the if already existed

surli replied:
Well actually since I call the method inside the if it's a bit redundant to check it again here... removing the check

manuelleduc commented on discussions-server/src/main/java/org/xwiki/contrib/discussions/script/DiscussionsScriptService.java:
```diff
@@ -226,6 +231,26 @@ public Message createMessage(String content, String syntax, String reference,
+     * @see MessageHolderReferenceService
+     */
+    @Unstable
+    public DocumentReference getNextMessageHolderReference(String serializedDiscussionReference,
```
Can't we make `serializedDiscussionReference` a `DocumentReference` and let Velocity do the resolution?

surli replied:
No a `serializedDiscussionReference` is not related to a `DocumentReference` it's a `DiscussionReference`. Right now I don't have the mechanism to perform automatic conversion in velocity and the whole script service API already uses String for those in other existing method. So I won't change that right now.

manuelleduc commented on discussions-store/discussions-store-default/src/main/java/org/xwiki/contrib/discussions/store/internal/DefaultMessageHolderReferenceService.java:
```diff
@@ -0,0 +1,114 @@
+            Map<DiscussionReference, DocumentReference> temporaryReferencesMapForCurrentSession =
+                this.getTemporaryReferencesMapForCurrentSession();
+            DocumentReference result;
+            if (!temporaryReferencesMapForCurrentSession.containsKey(discussionReference)) {
```
Can be shortened using `Map.putIfAbsent`

surli replied:
Indeed, I always forgot about this construct..

surli committed:
DISCUSSION-42: Upload of attachment is not properly supported

  * Provide some unit tests
  * Fix minor problems

manuelleduc commented on discussions-store/discussions-store-default/src/test/java/org/xwiki/contrib/discussions/store/internal/DefaultMessageHolderReferenceServiceTest.java:
```diff
@@ -0,0 +1,146 @@
+    @BeforeEach
+    void setup()
+    {
+        this.context = mock(XWikiContext.class);
```
Very much a question of style but I tend to use the `@Mock` annotation for mock fields.

surli replied:
I stopped doing it a while ago when having sometimes problems with order (e.g. when trying to use a mock in a beforeComponent call) so now I always use this to have better control.

surli committed:
DISCUSSION-42: Upload of attachment is not properly supported

  * Fix rendering of messages

surli committed:
DISCUSSION-42: Upload of attachment is not properly supported

Provides a dedicated API to allow creating a message from a request,
without using the discussion handler. The idea behind that work is that
in some cases we want to be able to create discussion message withing
another activity (in my case, I want to use that when performing a
change request review). And we don't necessarily want to perform two
different requests (one for the activity and one for the message
creation). So being able to create the message directly from the request
would be helpful.

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/server/DiscussionMessageRequestCreator.java:
```diff
@@ -0,0 +1,118 @@
+    String CONTENT_PARAMETER = "content";
+
+    /**
+     * Recommended request parameter defining the syntax of sent content. Default value is {@code XWiki/2.0} when the
```
Shouldn't it be `{@code xwiki/2.0}` (lower-cased)?

surli replied:
indeed, old habit of writing XWiki

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/server/DiscussionMessageRequestCreator.java:
```diff
@@ -0,0 +1,118 @@
+     * It <strong>might</strong> contain also those parameters:
+     * <ul>
+     *     <li>{@code content_syntax}: the syntax of the provided content. If the parameter is not sent, the content is
+     *     parsed as {@code XWiki/2.0}</li>
```
same here ;)

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/server/DiscussionServerException.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @param statusCode the http status code that should be answered
+     * @param msg the message of the exception
+     * @param e the original error
```
`e` is usually named `cause`

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/server/DiscussionServerException.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public int getStatusCode()
+    {
+        return statusCode;
```
Missing `this`

manuelleduc commented on discussions-server/src/main/java/org/xwiki/contrib/discussions/internal/server/DefaultDiscussionMessageRequestCreator.java:
```diff
@@ -0,0 +1,184 @@
+        if (StringUtils.isBlank(serializedReference)) {
+            throw new DiscussionServerException(HttpServletResponse.SC_BAD_REQUEST,
+                "The discussion reference has not been provided.");
+        } else {
```
Not sure the else is needed as you throw an exception in the if.

manuelleduc commented on discussions-server/src/main/java/org/xwiki/contrib/discussions/internal/server/DefaultDiscussionMessageRequestCreator.java:
```diff
@@ -0,0 +1,184 @@
+            DiscussionReference discussionReference =
+                this.discussionReferencesResolver.resolve(serializedReference, DiscussionReference.class);
+            Optional<Discussion> discussionOptional = this.discussionService.get(discussionReference);
+            if (!discussionOptional.isPresent()) {
```
[orElseThrow](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElseThrow-java.util.function.Supplier-)  is more idiomatic.

manuelleduc commented on discussions-server/src/main/java/org/xwiki/contrib/discussions/internal/server/DefaultDiscussionMessageRequestCreator.java:
```diff
@@ -0,0 +1,184 @@
+        return messageOptional;
+    }
+
+    private String getContent(HttpServletRequest request)
```
Can be simplified using `RequestParameterConverter` but only on XWiki 13.5RC1+, should we consider upgrading the parent version?

surli replied:
Well part of this PR is about upgrading the parent version to 14.10 :) Here it's because I just copy/pasted what was already existing, but I agree I should use `RequestParameterConverter` I'll check for it.

tmortagne replied:
From what I understand, this pull request also upgrades to 14.10.

tmortagne commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/server/DiscussionMessageRequestCreator.java:
```diff
@@ -0,0 +1,118 @@
+ * @version $Id$
+ * @since 2.1
+ */
+@Unstable
```
Not sure how to interpret that for a contrib extension in which you can keep the same major version for years (no concept of yearly cycle).

surli replied:
I'm using those for the contrib extension I contribute too using them with semver: I remove the unstable whenever I move to next major.

tmortagne replied:
> I remove the unstable whenever I move to next major.

That's what it means in XWiki Standard, but the major version change extremely rarely in most contrib extensions and keeping an API forbidden (which is pretty much what `@Unstable` means) for years does not make much sense.

surli replied:
I see your point, that's probably something to discuss on the forum since it's part of our best practices for XS and we're supposed to reuse them for contrib. Now I won't change that here since I already used that annotation in other places of the extension (when moving to 2.0), so I prefer to keep consistency for now at least. 
Also note that here Discussion is evolving a lot, so it kinds of make sense.

tmortagne replied:
> we're supposed to reuse them for contrib

When they make sense, but current `@Unstable` definition explicitly mention XWiki Standard cycles :)

tmortagne commented on discussions-server/src/main/java/org/xwiki/contrib/discussions/internal/server/DefaultDiscussionMessageRequestCreator.java:
```diff
@@ -0,0 +1,184 @@
+        // Handle temporary uploads
+        List<String> temporaryUploadedFiles = editForm.getTemporaryUploadedFiles();
+        parameters.put(TEMPORARY_UPLOADED_ATTACHMENTS, temporaryUploadedFiles);
+        Optional<Message> messageOptional;
```
Feels a bit strange to declare that here and assign it at the end of the method.

surli replied:
indeed, probably historical reasons  since this method has evolved a lot. I'll move it

surli committed:
DISCUSSION-42: Upload of attachment is not properly supported

  * Some changes following review from Thomas and Manuel
  * Rely on the RequestParameterConverter to perform content conversion
    instead of re-implementing it

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/domain/references/AbstractDiscussionReference.java:
```diff
@@ -34,7 +36,7 @@
  */
 @Unstable
-public abstract class AbstractDiscussionReference
+public abstract class AbstractDiscussionReference implements Serializable
```
Do we have a rule regarding Serializable and serialVersionUID?

surli replied:
I don't recall any
